src/util/mergeUpdates.ts

```diff
@@ -20,15 +20,36 @@ const mergeUpdates = <T>(
 
   Object.entries(mergee).forEach(([key, value]) => {
     if (value) {
-      // ignore pending objects that would overwrite non-pending objects by default
-      if (
-        overwritePending ||
-        !(value as MaybePending).pending ||
-        !mergeInto[key] ||
-        (mergeInto[key] as MaybePending).pending
-      ) {
+      const incomingPending = (value as MaybePending).pending
+      const currentPending = (mergeInto[key] as MaybePending)?.pending
+
+      // Do not overwrite an explicitly non-pending object with a pending object unless overwritePending is true.
+      // This prevents a thought that has finished loading (pending === false) from being set back to pending === true
+      // when the DataProvider returns the stale value that still has pending === true. See issue where pending thoughts
+      // remain pending indefinitely after navigating to them.
+
+      // Scenarios:
+      // 1. overwritePending is true: always allow overwrite (original behaviour).
+      // 2. currentPending === true: allow overwrite so that pending can be cleared.
+      // 3. currentPending === false && incomingPending === true: block overwrite to keep non-pending state.
+      // 4. currentPending === false && incomingPending === false/undefined: allow overwrite (other fields may update).
+
+      // Determine if the incoming object has its own pending property.
+      const hasPendingProp = Object.prototype.hasOwnProperty.call(value as object, 'pending')
```

Comment:
This would differentiate `{ pending: undefined }` from `{}`, which is not a good design. It's confusing enough to distinguish `{ pending: undefined }` from `{ pending: false }`.

Reply:
On second thought (assuming if we don't find a way to handle it in `fetchDescendants`), I think we should move the `clearPending` logic into `mergeUpdates` and make it thought-specific. The pending flag is already thought-specific in practice, despite the ostensibly generic way the function is currently written. The `pending` flag is unique to thoughts, so I think it's okay to build it into `mergeUpdates`.

Injecting `clearPending` is generic, but it's a pretty ugly interface. In this case I don't think it's worth the cost.